### PR TITLE
Feature/#47 action log structure

### PR DESCRIPTION
**Pull request by marcinkp**
Allows Nodes to add log data to processed Fragments. This data can be then visualized or used during debug sessions.
## Description

Graph node gets a fragment and processes it. It is described as a function:
`F -> (F', T)`

We want to extend this function by allowing node processing to add some logs. The new function can be described as:
`F -> (F', T, L)`

Each Node can produce some debug data that can be consumed by built-in or custom listeners. There are two levels: ERROR or INFO. The default one is ERROR. Each Node can produce its own debug data.

## Motivation and Context
Make graph processing easy to analyze.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Changelog
+All notable changes to `knotx-fragments` will be documented in this file.
+
+## Unreleased
+List of changes that are finished but not yet released in any final version.
+- [PR-45](https://github.com/Knotx/knotx-fragments/pull/45) - Action log mechanism implementation.
```
Comment (marcinkp) [inline CHANGELOG.md]: chnage PR number

File: handler/api/src/main/java/io/knotx/fragments/handler/api/domain/FragmentResult.java [hunk 2]
```diff
@@ -74,6 +87,15 @@ public String getTransition() {
     }
   }
 
+  /**
+   * Log produced by node execution.
+   *
+   * @return action log
```
Comment (marcinkp) [inline handler/api/src/main/java/io/knotx/fragments/handler/api/domain/FragmentResult.java]: rename

File: handler/api/src/main/java/io/knotx/fragments/handler/api/domain/FragmentResult.java [hunk 3]
```diff
@@ -84,19 +106,21 @@ public boolean equals(Object o) {
...
-    return Objects.hash(fragment, transition);
+    return Objects.hash(fragment, transition, nodeLog);
   }
 
   @Override
   public String toString() {
     return "FragmentResult{" +
         "fragment=" + fragment +
         ", transition='" + transition + '\'' +
+        ", actionLog=" + nodeLog +
```
Comment (marcinkp) [inline handler/api/src/main/java/io/knotx/fragments/handler/api/domain/FragmentResult.java]: generate

File: handler/README.md [hunk 4]
```diff
@@ -120,7 +120,20 @@ Knot.x provides two node implementations:
 - **Action node** that represents simple steps in a graph such as integration with a data source
 - **Subtasks node** that is a list of unnamed tasks (subtasks) that are evaluated in parallel
 
+####### Node Processing Log
```
Comment (tomaszmichalak) [inline handler/README.md]:
```suggestion
####### Fragment processing log
```

File: handler/README.md [hunk 5]
```diff
@@ -120,7 +120,20 @@ Knot.x provides two node implementations:
...
+
+Node processing produce flat logs in the following structure. 
+
+| Task | Node alias | Status | Transition | Node Log |
```
Comment (tomaszmichalak) [inline handler/README.md]:
```suggestion
| Task name | Node id | Transition | Fragment's status | Node Logs |
```

File: handler/README.md [hunk 6]
```diff
@@ -120,7 +120,20 @@ Knot.x provides two node implementations:
...
+|---|---|---|---|---|
+| TASK | A1 | SUCCESS | T1 | L-A1 |
```
Comment (tomaszmichalak) [inline handler/README.md]: Can you please some example here. You can reuse the image from engine that presents processing states.
```
A
|     \
B    C
```

File: handler/README.md [hunk 7]
```diff
@@ -196,6 +209,16 @@ two independent tasks (graphs) with one node (action).
 See the [example section](#the-example) for a more complex scenario. Before we see the full 
 power of graphs, we need to understand how nodes are connected.
 
+####### Subtasks Processing Log
```
Comment (tomaszmichalak) [inline handler/README.md]:
```suggestion
####### Fragment processing log entries
```

File: handler/README.md (hunk 6 above)
Comment (tomaszmichalak) [reply handler/README.md]: You can also use `_success` transitions to be more accurate. You can describe also what is in the log when operation ends correctly or raises some exception.

File: handler/README.md [hunk 8]
```diff
@@ -196,6 +209,16 @@ two independent tasks (graphs) with one node (action).
...
+
+Let's consider above subtasks definition. The processing log will look as follows.  
+
+| Task | Node alias | Status | Transition | Node Log |
+|---|---|---|---|---|
+| TASK | book-rest-api | SUCCESS | T1 | L-book-rest-api |
```
Comment (tomaszmichalak) [inline handler/README.md]:
```suggestion
| TASK | book-rest-api | SUCCESS | _success| L-book-rest-api |
```

File: handler/README.md [hunk 9]
```diff
@@ -196,6 +209,16 @@ two independent tasks (graphs) with one node (action).
 See the [example section](#the-example) for a more complex scenario. Before we see the full 
 power of graphs, we need to understand how nodes are connected.
 
+####### Subtasks Processing Log
+
+Let's consider above subtasks definition. The processing log will look as follows.  
+
+| Task | Node alias | Status | Transition | Node Log |
```
Comment (tomaszmichalak) [inline handler/README.md]: Please update the column order.

File: handler/README.md [hunk 10]
```diff
@@ -196,6 +209,16 @@ two independent tasks (graphs) with one node (action).
...
+|---|---|---|---|---|
+| TASK | book-rest-api | SUCCESS | T1 | L-book-rest-api |
+| TASK | author-rest-api | SUCCESS | T2 | L-author-rest-api |
+| TASK | composite | SUCCESS | T3 |  |
```
Comment (tomaszmichalak) [inline handler/README.md]: Can you please describe what is in log when we have nested composite nodes and if an error occurs.

File: handler/README.md [hunk 11]
```diff
@@ -416,3 +442,26 @@ doAction = product-cb
 Please note that cacheKey can be parametrized with request data like params, headers etc. Read 
 [Knot.x HTTP Server Common Placeholders](https://github.com/Knotx/knotx-server-http/tree/master/common/placeholders)
 documentation for more details.
+
+### Actions log
+
+Each node can produce specific execution logs. As mention before those logs are stored in  `Node processing log`
```
Comment (tomaszmichalak) [inline handler/README.md]: Grammarly?

File: handler/README.md (hunk 10 above)
Comment (tomaszmichalak) [reply handler/README.md]: Subgraphs are evaluated, then at the end composite node id entry is added...

File: handler/README.md [hunk 12]
```diff
@@ -416,3 +442,26 @@ doAction = product-cb
 Please note that cacheKey can be parametrized with request data like params, headers etc. Read 
 [Knot.x HTTP Server Common Placeholders](https://github.com/Knotx/knotx-server-http/tree/master/common/placeholders)
 documentation for more details.
+
+### Actions log
```
Comment (tomaszmichalak) [inline handler/README.md]:
```suggestion
### Action's node log structure
```

File: handler/README.md [hunk 13]
```diff
@@ -416,3 +442,26 @@ doAction = product-cb
...
+  _logs: {}
+  _doAction [
+    AL, AL, ...
+  ]
+}
+```
+
+where:
+ A1 is `behavior` 
+ A1`, A1`` and A1 ``` are corresponding execution of wrapped simple action
```
Comment (tomaszmichalak) [inline handler/README.md]: ?

File: handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineCompositeNodeTest.java [hunk 14]
```diff
@@ -43,18 +58,9 @@
...
-import java.util.Map;
-import java.util.concurrent.TimeUnit;
-import java.util.function.Consumer;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.extension.ExtendWith;
 
 @ExtendWith(VertxExtension.class)
+// TODO add tests
```
Comment (tomaszmichalak) [inline handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineCompositeNodeTest.java]: ?

File: handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineSingleOperationTest.java [hunk 15]
```diff
@@ -274,6 +276,48 @@ void expectErrorAndSuccessEventLogEntries(VertxTestContext testContext, Vertx ve
         ));
   }
 
+  @Test
+  @DisplayName("Expect node debug in log event log entries when success transition handled.")
```
Comment (tomaszmichalak) [inline handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineSingleOperationTest.java]: Can you please add some mock behaviour and prepare test for it.
- behaviour that does 3 retries (failing action)

Expected result:
node log : behaviour log + [first invovation, second invocation, third invocation]

File: handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineCompositeNodeTest.java [hunk 16]
```diff
@@ -111,9 +118,12 @@ void expectSuccessWhenSingleProcessingEnds(VertxTestContext testContext, Vertx v
   void expectSuccessEventLogEntry(VertxTestContext testContext, Vertx vertx)
       throws Throwable {
     // given
+    JsonObject successActionLog = new JsonObject().put("debug", "success");
```
Comment (marcinkp) [inline handler/engine/src/test/java/io/knotx/fragments/engine/TaskEngineCompositeNodeTest.java]: two different logs

File: handler/README.md [hunk 17]
```diff
@@ -95,16 +95,15 @@ There are two sections:
 - `node` defines a fragment processing logic
 - `onTransitions` is a map that represents outgoing edges in a graph
 
-##### Node processing
-The node responsibility can be described as: 
-> Graph node gets a fragment, processes it and responds with Transition. So a node is the function 
->`F -> (F', T)` where `F` is the Fragment, `F'` is a modified Fragment and `T` is the Transition.
+##### Node
+Node's definition is are described [here](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#node).
```
Comment (Skejven) [inline handler/README.md]:
```suggestion
Node is described [here](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#node).
```

File: handler/README.md [hunk 18]
```diff
@@ -95,16 +95,15 @@ There are two sections:
...
 
-The node definition is abstract. It allows to define simple processing nodes but also more complex 
-structures such as a list of subgraphs. Furthermore, such a definition inspires to provide custom 
-node implementations.
+Fragments Handler introduces defines custom node types that are finally converted to the
```
Comment (Skejven) [inline handler/README.md]:
```suggestion
Fragments Handler introduces custom node types that are finally converted to the 
```

File: handler/README.md [hunk 19]
```diff
@@ -95,16 +95,15 @@ There are two sections:
...
->`F -> (F', T)` where `F` is the Fragment, `F'` is a modified Fragment and `T` is the Transition.
+##### Node
+Node's definition is are described [here](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#node).
 
-The node definition is abstract. It allows to define simple processing nodes but also more complex 
-structures such as a list of subgraphs. Furthermore, such a definition inspires to provide custom 
-node implementations.
+Fragments Handler introduces defines custom node types that are finally converted to the 
+[engine node types](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine#node-types).
+It allows to quickly introduce new node types, with different configuration options, without modifying
```
Comment (Skejven) [inline handler/README.md]:
```suggestion
It allows to quickly add new node types, with different configuration options, without modifying
```

File: handler/README.md [hunk 20]
```diff
@@ -95,16 +95,15 @@ There are two sections:
...
+the engine.
 
-The `node` configuration is simple:
+Each node defines its custom factory. The configuration is simple::
```
Comment (Skejven) [inline handler/README.md]:
```suggestion
Each node defines its custom factory. The configuration is simple:
```

File: handler/engine/README.md [hunk 21]
```diff
@@ -8,19 +8,25 @@ Any *Fragment* can define its processing path - a **Task** (which is a **directe
 A **Task** specifies the nodes through which Fragments will be routed by the Task Engine. 
 Each Node may define possible *outgoing edges* - **Transitions**.
 
-### Node
+## Node
 The node responsibility can be described as: 
-> Graph node gets a fragment, processes it and responds with Transition. So a node is the function 
->`F -> (F', T)` where `F` is the Fragment, `F'` is a modified Fragment and `T` is the Transition.
+> Graph node gets a fragment, processes it, add some processing logs and responds with a transition.
```
Comment (Mateusz512) [inline handler/engine/README.md]:
```suggestion
> Graph node gets a fragment, processes it, adds some processing logs and responds with a transition. 
```

File: handler/engine/README.md [hunk 22]
```diff
@@ -108,4 +118,31 @@ The images below illustrates the above rules.
...
+- node identifier
+- node status
+- [node log](#node-log)
+- transition
+- timestamp.
+
+Node status is a simple text value managed by the engine. It resembles a fragment's status but is a 
+bit more accurate (such as a `UNSUPPORTED_TRANSITION` value).
+
+Let's see the example fragment's log. There is a fragment that defines a task named `taskName`. The
```
Comment (Mateusz512) [inline handler/engine/README.md]:
```suggestion
Let's see an example fragment's log. There is a fragment that defines a task named `taskName`. The 
```

File: handler/engine/README.md [hunk 23]
```diff
@@ -108,4 +118,31 @@ The images below illustrates the above rules.
...
+
+![A and B ends correctly](assets/images/a_success_b_success.png)
+
+The `A` node responds with the `_success` transition. Then the `B` node starts processing and responds 
+with the `_succcess` transition. Finally, the fragment status is `SUCCESS` and the fragment's log contains:
+
+| Task       | Node identifier | Node status | Transition | Node Log        |
+|------------|-----------------|-------------|------------|-----------------|
+| `taskName` | `A`             | SUCCESS     | `_success` | { A node log }  |
+| `taskName` | `B`             | SUCCESS     | `_success` | { A node log }  |
```
Comment (Mateusz512) [inline handler/engine/README.md]:
```suggestion
| `taskName` | `B`             | SUCCESS     | `_success` | { B node log }  |
```

File: handler/engine/README.md [hunk 24]
```diff
@@ -108,4 +118,31 @@ The images below illustrates the above rules.
...
+
+Node status is a simple text value managed by the engine. It resembles a fragment's status but is a 
+bit more accurate (such as a `UNSUPPORTED_TRANSITION` value).
+
+Let's see the example fragment's log. There is a fragment that defines a task named `taskName`. The 
+task is a graph of two nodes: `A` and `B`.
+
+![A and B ends correctly](assets/images/a_success_b_success.png)
+
+The `A` node responds with the `_success` transition. Then the `B` node starts processing and responds
```
Comment (Mateusz512) [inline handler/engine/README.md]: I think we're missing some punch line in this example, it could be more verbose + what is the outcome for the reader?

Comment (tomaszmichalak) [reply handler/engine/README.md]: I would suggest something during actions updates.

**Pull request merged**